### PR TITLE
[update] update container_kill chaos status

### PR DESCRIPTION
**Pull request by fewdan**
### What problem does this PR solve?
If at least one pod does not have a specified container in the target pods, the state of chaos becomes failed.

### What is changed and how does it work?
If at least one pod does not have a specified container in the target pods, return the error info.

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility

Related changes

- [ ] Need to update the documentation

### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: controllers/podchaos/containerkill/types.go [hunk 1]
```diff
@@ -96,6 +98,10 @@ func (r *endpoint) Apply(ctx context.Context, req ctrl.Request, obj v1alpha1.Inn
 		return err
 	}
 
+	if allHaveContainer == false {
+		return fmt.Errorf("at least one pod contains no designated container")
+	}
+
 	podchaos.Status.Experiment.PodRecords = make([]v1alpha1.PodStatus, 0, len(pods))
 	for _, pod := range pods {
```
Comment (fewdan) [inline controllers/podchaos/containerkill/types.go:106]: I have a doubt: if some pods have a container and some pods do not have a container in the selected pods, at this time, if the above returns directly, the corresponding pod_records will not be recorded. Is this reasonable? Or should you record pod_records and then "return fmt.Errorf("at least one pod contains no designated container")"?

File: controllers/podchaos/containerkill/types.go [hunk 2]
```diff
@@ -88,6 +89,7 @@ func (r *endpoint) Apply(ctx context.Context, req ctrl.Request, obj v1alpha1.Inn
 		}
 
 		if haveContainer == false {
+			allHaveContainer = false
 			r.Log.Error(nil, fmt.Sprintf("the pod %s doesn't have container %s", pod.Name, podchaos.Spec.ContainerName))
```
Comment (cwen0) [inline controllers/podchaos/containerkill/types.go:93]: We should return this error here

File: controllers/podchaos/containerkill/types.go (hunk 1 above)
Comment (cwen0) [reply controllers/podchaos/containerkill/types.go:106]: Yes, I think we should record this pod_records.

File: controllers/podchaos/containerkill/types.go (hunk 2 above)
Comment (fewdan) [reply controllers/podchaos/containerkill/types.go:93]: @cwen0
Then there will be a situation that is, if we have pod1, pod2, and there is a specified container in pod2, if we traverse pod1 first, and then return an error directly here, it is possible that pod2 is not processed.

**Comment by STRRL** [comment]
Hi @fewdan, I prefer to complete the checking of `allHaveContainers` **before killing some containers**.

Current implementation will bring a mess, for example:

I have 2 pods that have two containers so-called "web-app" and "monitoring",  and we have another 3 pods that have two containers so-called "another-web-app" and "monitoring". All 5 pods have one label "app: web".

Then I publish a container kill chaos, which wants to kill the "web-app" container, with labelSelector "app: web".

Obviously, the last 3 pods do not contain the container called "web-app", so this chaos experiment must be failed.

How do we resolve the first 2 pods which indeed contain the "web-app" container? The current implementation will kill these containers, and the Chaos Experiment falls to failure, emphasize again, the chaos failed, but something bad (the containers have been killed) already happened.

I prefer to make a check at first, if any pods mismatch with the required container, the chaos experiment goes failed, and no containers will be killed

**Comment by cwen0** [comment]
> Hi @fewdan, I prefer to complete the checking of `allHaveContainers` **before killing some containers**.
> 
> Current implementation will bring a mess, for example:
> 
> I have 2 pods that have two containers so-called "web-app" and "monitoring", and we have another 3 pods that have two containers so-called "another-web-app" and "monitoring". All 5 pods have one label "app: web".
> 
> Then I publish a container kill chaos, which wants to kill the "web-app" container, with labelSelector "app: web".
> 
> Obviously, the last 3 pods do not contain the container called "web-app", so this chaos experiment must be failed.
> 
> How do we resolve the first 2 pods which indeed contain the "web-app" container? The current implementation will kill these containers, and the Chaos Experiment falls to failure, emphasize again, the chaos failed, but something bad (the containers have been killed) already happened.
> 
> I prefer to make a check at first, if any pods mismatch with the required container, the chaos experiment goes failed, and no containers will be killed

I agree with this idea.

**Comment by fewdan** [comment]
> Hi @fewdan, I prefer to complete the checking of `allHaveContainers` **before killing some containers**.
> 
> Current implementation will bring a mess, for example:
> 
> I have 2 pods that have two containers so-called "web-app" and "monitoring", and we have another 3 pods that have two containers so-called "another-web-app" and "monitoring". All 5 pods have one label "app: web".
> 
> Then I publish a container kill chaos, which wants to kill the "web-app" container, with labelSelector "app: web".
> 
> Obviously, the last 3 pods do not contain the container called "web-app", so this chaos experiment must be failed.
> 
> How do we resolve the first 2 pods which indeed contain the "web-app" container? The current implementation will kill these containers, and the Chaos Experiment falls to failure, emphasize again, the chaos failed, but something bad (the containers have been killed) already happened.
> 
> I prefer to make a check at first, if any pods mismatch with the required container, the chaos experiment goes failed, and no containers will be killed

ok, I think it's a good idea, i'll modify it.

**Review by cwen0 — APPROVED** [review]
LGTM

File: controllers/podchaos/containerkill/types.go [hunk 3]
```diff
@@ -64,17 +64,38 @@ func (r *endpoint) Apply(ctx context.Context, req ctrl.Request, obj v1alpha1.Inn
...
 
+		for containerIndex := range pod.Status.ContainerStatuses {
+			containerName := pod.Status.ContainerStatuses[containerIndex].Name
+			if containerName == podchaos.Spec.ContainerName {
+				haveContainer = true
+				break
+			}
+		}
+
+		if haveContainer == false {
```
Comment (STRRL) [inline controllers/podchaos/containerkill/types.go:80]:
```suggestion
		if !haveContainer {
```

File: controllers/podchaos/containerkill/types.go [hunk 4]
```diff
@@ -64,17 +64,38 @@ func (r *endpoint) Apply(ctx context.Context, req ctrl.Request, obj v1alpha1.Inn
...
 		pod := &pods[podIndex]
 		haveContainer := false
 
+		for containerIndex := range pod.Status.ContainerStatuses {
+			containerName := pod.Status.ContainerStatuses[containerIndex].Name
+			if containerName == podchaos.Spec.ContainerName {
+				haveContainer = true
+				break
+			}
+		}
```
Comment (STRRL) [inline controllers/podchaos/containerkill/types.go:78]: I prefer that:
```suggestion
		for _, item := range pod.Status.ContainerStatuses {
			if item.Name == podchaos.Spec.ContainerName {
				haveContainer = true
				break
			}
		}
```

File: controllers/podchaos/containerkill/types.go [hunk 5]
```diff
@@ -64,17 +64,38 @@ func (r *endpoint) Apply(ctx context.Context, req ctrl.Request, obj v1alpha1.Inn
 		return err
 	}
 
-	g := errgroup.Group{}
+	podsNotHaveContainer := []string{}
```
Comment (STRRL) [inline controllers/podchaos/containerkill/types.go:67]:
```suggestion
	var podsNotHaveContainer []string
```

**Review by STRRL — CHANGES_REQUESTED** [review]
@fewdan PTAL

File: controllers/podchaos/containerkill/types.go (hunk 5 above)
Comment (fewdan) [reply controllers/podchaos/containerkill/types.go:67]: Okay, let me revise the above question.

**Review by STRRL — APPROVED** [review]
LGTM

**Review by cwen0 — APPROVED** [review]

**Pull request merged**